File: p0/p14/LongestCommonPrefix.java
package leetcode.p0.p14;

public class LongestCommonPrefix {
    public static String longestCommonPrefix1(String[] strs) {
        String commonPrefix = "";
        for (int i = 0; i < strs[0].length(); i++) {
            char c = strs[0].charAt(i);
            for (String str : strs) {
                if (i >= str.length() || c != str.charAt(i)) {
                    return commonPrefix;
                }
            }
            commonPrefix += c;
        }
        return commonPrefix;
    }

    public static String longestCommonPrefix(String[] strs) {
        String s = "";
        if (strs.length == 1) {
            return strs[0];
        }
        for (int i = 0; i < strs.length - 1; i++) {
            if (i == 0) {
                s = commonPrefix(strs[0], strs[1]);
            } else {
                s = commonPrefix(s, strs[i + 1]);
            }
        }
        return s;
    }

    private static String commonPrefix(String s1, String s2) {
        String prefix = "";
        for (int i = 0; i < Math.min(s1.length(), s2.length()); i++) {
            if (s1.charAt(i) == s2.charAt(i)) {
                prefix += s1.charAt(i);
            } else {
                return prefix;
            }
        }
        return prefix;
    }


    public static void main(String[] args) {
        String[] strs = new String[]{"flower", "flow", "flight"};
        System.out.println(longestCommonPrefix(strs));
    }
}

File: p0/p14/longest_common_prefix.py
"""
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".



Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"
Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.


Constraints:

1 <= strs.length <= 200
0 <= strs[i].length <= 200
strs[i] consists of only lowercase English letters if it is non-empty.
"""
from typing import List


class Solution:
    """
    Jun 24, 2025 13:23
    Time Complexity: O(n)
    Space Complexity: O(1)
    """

    def longestCommonPrefix(self, strs: List[str]) -> str:
        result = ""
        i = 0
        c = ""
        while True:
            for s in strs:
                if i >= len(s):
                    return result
                if not c:
                    c = s[i]
                else:
                    if s[i] != c:
                        return result
            result += c
            c = ""
            i += 1


if __name__ == '__main__':
    sol = Solution()
    strs = ["dog", "racecar", "car"]
    print(sol.longestCommonPrefix(strs))
    strs = ["flower", "flow", "flight"]
    print(sol.longestCommonPrefix(strs))
